### DISCUSSION TITLE
Is there a way to inject dependencies into server actions?

### DISCUSSION THREAD

**Discussion by ErfanEbrahimnia** [vercel/next.js#68984] [Help]
### Summary

I'm working on a package and would like to offer the following API:

```jsx
const Component = create({
  databaseAdapter: {
    async save() {
      console.log("saved to db");
    },
  },
});

export default function Page() {
  return <Component />;
}
```

The intended implementation of create is as follows:

```jsx
export function create({ databaseAdapter }) {
  async function someAction(formData) {
    "use server";

    const data = formData.get("data");
    await databaseAdapter.save(data);
  }

  return function Component() {
    return (
      <form action={someAction}>
        <button type="submit">Submit</button>
      </form>
    );
  };
}
```

Essentially, I'm trying to inject the database adapter into the server action. However, the current code doesn’t work as expected. Next.js attempts to serialize databaseAdapter.save, resulting in this error:

_Error: Functions cannot be passed directly to Client Components unless you explicitly expose it by marking it with "use server". Or maybe you meant to call this function rather than return it._

Adding "use server" to the save method doesn’t resolve the issue either. Is there a way to implement this as I intend?

**Comment by bestickley** [comment]
@ErfanEbrahimnia, I'm running into this same issue of trying to pass a function (non-serializable) to a server action but running the error:

```
Error: Functions cannot be passed directly to Client Components unless you explicitly expose it by marking it with "use server". Or maybe you meant to call this function rather than return it.
```

I haven't found anything in the docs to say I cannot do this. If you CMD + F "serializable" you'll see it's always referencing return values or parameters, never closure / scoped variables.

**Reply by ErfanEbrahimnia** [reply]
It seems you can't use Server Actions in that way, so I switched to a catch-all endpoint approach instead.

**Reply by bestickley** [reply]
@ErfanEbrahimnia, so to summarize: we cannot pass in a function as props to a server component and then reference/call that in a server action. Does that sound right to you? This needs to be documented.

**Reply by bestickley** [reply]
It doesn't make sense to me why there are limitations on this. I thought it didn't need to serializable because it's all on the server. You can pass the reference to the function into a client component and it would be called but references in the server action closure shouldn't matter because it stays on server.

**Reply by ErfanEbrahimnia** [reply]
I'm not entirely sure, but I think the bundler might not be sophisticated enough to differentiate between server action dependencies coming from the server versus the client

**Reply by bestickley** [reply]
@shuding , are you able to shed any light here? Saw you tagged in another server actions issue. Here is example repo: https://stackblitz.com/edit/stackblitz-starters-tpkzew?file=package.json

**Reply by shuding** [reply]
> Adding "use server" to the save method doesn’t resolve the issue either

This feels like a bug — I'll take a look!

**Reply by ErfanEbrahimnia** [reply]
Any updates, @shuding? If this is a bug, I'll hold off on releasing my package until the fix is available
